### PR TITLE
fix: Add alert for incoming messages on dashboard

### PR DESCRIPTION
**Pull request by nithindavid**
Fixes - #1345

**Note**
- Does not work on safari, Seems to work on Firefox/Chrome. (User needs to make some interaction with page, scroll, touch )
- Need to add more cases to when to play the sound.
- Need an initial review on sound part.

### CODE REVIEW

File: app/views/layouts/vueapp.html.erb [hunk 1]
```diff
@@ -50,5 +50,8 @@
     <div id="app"></div>
     <noscript id="noscript">This app works best with JavaScript enabled.</noscript>
     <%= yield %>
+    <audio autoplay loop muted id="notif-audio">
+      <source src="<%= audio_path 'ding.mp3' %>" type="audio/mp3">
```
Comment (pranavrajs) [inline app/views/layouts/vueapp.html.erb:54]: Why can't we use the existing audio file?

**Review by pranavrajs — COMMENTED** [review]
Except for the audio path, PR looks good. Can we add a setting to disable the sound notifications on UI, otherwise this would create a mess.

File: app/javascript/shared/helpers/AudioNotificationHelper.js [hunk 2]
```diff
@@ -7,3 +7,31 @@ export const playNotificationAudio = () => {
...
+
+export const newMessageNotification = data => {
+  if (document.hidden) {
+    playAudioFromIframe();
+  } else {
+    const { conversation_id: currentPageId } = window.WOOT.$route.params;
+    const { conversation_id: incomingConvId } = data;
+    if (currentPageId !== incomingConvId) {
+      playAudioFromIframe();
+    }
```
Comment (pranavrajs) [inline app/javascript/shared/helpers/AudioNotificationHelper.js:35]: Do we need to do this? I guess we should only alert the user with the sound only if the window is out of focus? 

@sojan-official thoughts?

**Comment by whande1992** [comment]
I also felt the need for sound notification, in addition to the visual one, as the agents cannot identify when there is a customer in line, a new conversation or message. I look forward to the appeal.

File: app/javascript/shared/helpers/AudioNotificationHelper.js (hunk 2 above)
Comment (sojan-official) [reply app/javascript/shared/helpers/AudioNotificationHelper.js:35]: The ideal case is if there is no user interaction on the page for x minutes. I could be away from my desk to get my coffee and definitely the sound would help me not to miss out the conversation.

If we have the UI for disabling audio alert I would go for sound on all notifications even when the page is on focus

- Accounts with high volume conversations can have the sounds disabled.
- Accounts where volume is low, the notification would help me not to miss out on messages. 

We could take it as a second iteration to disable the notifications if the user is active on the page

File: app/views/layouts/vueapp.html.erb (hunk 1 above)
Comment (nithindavid) [reply app/views/layouts/vueapp.html.erb:54]: Needs to be in assets folder to include in `.erb` files. @sojan-official Any idea to use the same file in `app/javascript/shared/assets/audio/ding.mp3` in `.erb` files?

**Comment by sojan-official** [comment]
Let's get this merged with a setting to disable audio alerts in the profile.
cc: @pranavrajs @nithindavid

File: app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue [hunk 3]
```diff
@@ -1,6 +1,32 @@
...
+        <p>
+          {{ $t('PROFILE_SETTINGS.FORM.AUDIO_NOTIFICATIONS_SECTION.NOTE') }}
+        </p>
+      </div>
+      <div class="columns small-9">
+        <div>
+          <input
+            id="audio_enable_alert"
+            v-model="enableAudioNotifications"
+            class="notification--checkbox"
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue:17]: unused classname

File: app/views/layouts/vueapp.html.erb (hunk 1 above)
Comment (pranavrajs) [reply app/views/layouts/vueapp.html.erb:59]: Please move it to the public folder in this case.

Comment (pranavrajs) [reply app/views/layouts/vueapp.html.erb:59]: Use an .ogg file not mp3

File: config/environments/production.rb [hunk 4]
```diff
@@ -76,7 +76,7 @@
   # Use a different logger for distributed setups.
   # require 'syslog/logger'
   # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
-  
+
```
Comment (pranavrajs) [inline config/environments/production.rb:79]: Unnecessary change, please revert.

File: app/javascript/shared/helpers/AudioNotificationHelper.js [hunk 5]
```diff
@@ -4,6 +4,44 @@ export const playNotificationAudio = () => {
...
+  } else {
+    const { conversation_id: currentPageId } = window.WOOT.$route.params;
+    const currentUserId = window.WOOT.$store.getters.getCurrentUserID;
+    const {
+      enable_audio_alerts: enableAudioNotifications = false,
+    } = window.WOOT.$store.getters.getUISettings;
+    const { conversation_id: incomingConvId, sender_id: senderId } = data;
+    const isFromCurrentUser = currentUserId === senderId;
+    const playAudio =
+      currentPageId !== incomingConvId &&
```
Comment (pranavrajs) [inline app/javascript/shared/helpers/AudioNotificationHelper.js:39]: `currentPageId` is confusing, please rename it to `currentConversationId`

File: app/javascript/shared/helpers/AudioNotificationHelper.js [hunk 6]
```diff
@@ -4,6 +4,44 @@ export const playNotificationAudio = () => {
...
+};
+
+export const newMessageNotification = data => {
+  if (document.hidden) {
+    playAudioFromIframe();
+  } else {
+    const { conversation_id: currentPageId } = window.WOOT.$route.params;
+    const currentUserId = window.WOOT.$store.getters.getCurrentUserID;
+    const {
+      enable_audio_alerts: enableAudioNotifications = false,
```
Comment (pranavrajs) [inline app/javascript/shared/helpers/AudioNotificationHelper.js:34]: `enableAudioAlerts`, please be consistent with the names

File: app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue [hunk 7]
```diff
@@ -277,6 +310,12 @@ export default {
 
       this.updateNotificationSettings();
     },
+    handleAudioInput(e) {
+      this.enableAudioNotifications = e.target.checked;
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue:314]: Please use either `enableAudioAlert` or `enableAudioNotification`, don't mix both.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please see the inline comments.

File: app/javascript/shared/helpers/AudioNotificationHelper.js [hunk 8]
```diff
@@ -4,6 +4,44 @@ export const playNotificationAudio = () => {
...
+    } = window.WOOT.$store.getters.getUISettings;
+    const { conversation_id: incomingConvId, sender_id: senderId } = data;
+    const isFromCurrentUser = currentUserId === senderId;
+    const playAudio =
+      currentPageId !== incomingConvId &&
+      !isFromCurrentUser &&
+      enableAudioNotifications;
```
Comment (pranavrajs) [inline app/javascript/shared/helpers/AudioNotificationHelper.js:41]: I think the audio alert should only be sent for incoming messages, if a teammate replies to the message why should we alert another user?

@sojan-official Thoughts

**Comment by pranavrajs** [comment]
A couple of other issues: 

1. There is no separation between the sections

Screenshot 2021-02-28 at 8 14 00 PM

2. Clicking on "Play audio notification" does not give me feedback. While other checkboxes show an alert that the settings have been updated.

3. Sound does not work in Firefox.

File: app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue (hunk 3 above)
Comment (nithindavid) [reply app/javascript/dashboard/routes/dashboard/settings/profile/NotificationSettings.vue:17]: Already in use classname

**Comment by nithindavid** [comment]
> While other checkboxes show an alert that the settings have been updated, clicking on "Play audio notification" does not give me feedback.

The Update UI doesn't have promise support, Which is out of scope of this issue. Will fix in a new PR since couple of other parts also needs to be updated.

**Comment by sojan-official** [comment]
- Go to settings page and enable audio alert
- go to conversations page and refresh the page 
- Go back to the settings page and the audio alert is unchecked ( refreshing the pages brings back the checkbox as checked )

Screenshot 2021-03-02 at 8 20 52 PM

**Comment by nithindavid** [comment]
> * Go to settings page and enable audio alert
> * go to conversations page and refresh the page
> * Go back to the settings page and the audio alert is unchecked ( refreshing the pages brings back the checkbox as checked )
> 
> Screenshot 2021-03-02 at 8 20 52 PM

Fixed

**Review by sojan-official — APPROVED** [review]
